.travis/check-ABI-API-compliance.sh

```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
+
+set -e
+
+function print_abi_api_breakages() {
+   echo "ABI breakages detected:"
+   cat compat_reports/libtango/${old_revision}_to_${new_revision}/abi_affected.txt | c++filt
+   echo "API breakages detected:"
+   cat compat_reports/libtango/${old_revision}_to_${new_revision}/src_affected.txt | c++filt
+   exit 1
+}
+
+# $1 string prefix
+# $2 git revision
+function generate_info() {
+  local prefix=$1
+  local revision=$2
+
+  git worktree remove $prefix-branch || true
+  git -c advice.detachedHead=false worktree add ${prefix}-branch ${revision}
+  mkdir ${prefix}-branch/build
+  cd ${prefix}-branch/build
+  cmake -DCMAKE_BUILD_TYPE=Debug -DTANGO_ABI_API_CHECKING=ON -DTANGO_COMPILE_TEST_SUITE=OFF ..
+  make
+  abi-dumper libtango.so -o ${base}/libtango-${prefix}-${revision}.dump -lver ${revision}
+  cd "${base}"
+}
+
+new_revision=$(git rev-parse HEAD)
+
+# old_revision is the first parent commit not in the history of this branch
+# see the help of git merge-base for the gritty detailed explanation
+old_revision=$(git merge-base --fork-point tango-9-lts HEAD)
```

Comment:
Aren't we interested more in whether the current branch is binary compatible with the PR target branch? I mean maybe we should trigger this job only for PRs and only for those those targeting branch 9.3-backports.

Travis sets some environmental variables that allow to determine whether this is a PR build and what is the name of the target branch.

Reply:
@mliszcz 

> Aren't we interested more in whether the current branch is binary compatible with the PR target branch? 

Yes we are exactly interested in that. As I switched now to github actions, I can't use the travis environment variable. I also would like the script to be runnable locally as well.

I think it is nice to have it for tango-9-lts as well just to teach us how compliance really works. The check is non-fatal only when targeting tango-9-lts, it will be fatal when I do the backport for 9.3-backports.

Reply:
Found it. Github action also allows to query the target branch. See `base_ref` in contexts, https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context.

Reply:
I was thinking about different thing. Consider below example:
```
      merge-base   tango-9-lts   refs/pull/ID/merge
          |            |                |
          V            V                V
    a1----a2-----------a3---------------a4
           \                            /
            b1-------------------------b2
                                       A
                                       |
                                 PR branch HEAD
```
Now we're checking that b2 (PR HEAD) is binary compatible with a2 (merge base). I think that for user (who is using latest version of target branch, that is a3, which may not even be binary compatible with a2) it is more interesting if a4 will still be compatible after we merge the PR.

We should be able to extract the merged ref from action trigger, see [GITHUB_REF here](https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-event-pull_request), or maybe something like `refs/pull/${{ github.event.number }}/merge` would work. I've also seen `github.event.after` used as a ref [here](https://github.com/actions/checkout/pull/102), although I do not know how it works.

On the other hand, what you currently did is also useful as it provides information whether the current branch introduces ABI breaking changes. I have no strong opinion on which of these two options is better.

Reply:
Ideally, we want to check the binary compatibility with a previous official release.
So we want to check if a4 (result of the merge) is binary compatible with the latest release tag we want to check (9.3.4-rc6 or even 9.3.3 right now, but might change in the future).

Reply:
@mliszcz Thanks, now I understand. I wanted the script also to locally work, so I avoided the additional headache of working with the merged result.

@bourtemb Comparing with a tagged release is also something we would like to know yes.